### PR TITLE
Vmray-rebuild

### PR DESCRIPTION
**Pull request by jochman**
## Status
In Progress (fixing bugs)

## TODO:
 fix vmray scripts to not break BC

## Related Issues
fixes: https://github.com/demisto/etc/issues/16079

## Description
added bunch of command and detonate file playbook

## Screenshots
Paste here any images that will help the reviewer

## Required version of Demisto
x.x.x

## Does it break backward compatibility?
   - No
        - But if someone has created custom script, it'll

## Must have
- [x] Tests
- [x] Documentation (with link to it) https://github.com/demisto/etc/issues/16519
- [x] Code Review

## Additional changes
Reverts the docker to basic demisto/python docker

### CODE REVIEW

File: Integrations/VMRay/integration-VMRay.py [hunk 1]
```diff
@@ -0,0 +1,636 @@
+""" GLOBAL PARAMS """
+API_KEY = demisto.params()["api_key"]
+SERVER = (
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove parenthesis

File: Integrations/VMRay/integration-VMRay.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+    del os.environ["https_proxy"]
+
+""" HELPER DICTS """
+SEVERITY_DICT = {
+    "malicious": "Malicious",
+    "suspicious": "Suspicious",
+    "not_suspicious": "Good",
+    "blacklisted": "Blacklisted",
+    "whitelisted": "Whitelisted",
+    "unknown": "Unknown",
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+}
+
+DBOTSCORE = {
+    "Malicious": 3,
+    "Suspicious": 2,
+    "Good": 1,
+    "Blacklisted": 3,
+    "Whitelisted": 1,
+    "Unknown": 0,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+
+    url = SERVER + url_suffix
+    r = requests.request(
+        method,
+        url,
+        json=body,
+        params=params,
+        headers=HEADERS,
+        files=files,
+        verify=USE_SSL,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+    )
+    if r.status_code not in {200, 201, 202, 204}:
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: are all these status codes needed ? can they all be returned ? are they all legit?

File: Integrations/VMRay/integration-VMRay.py [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+        error = str()
+        try:
+            js = r.json()
+            error += js.get("error_msg")
+        except ValueError:
+            error = r.text
+        return_error(
+            "Error in API call to VMRay [{}] - {}".format(r.status_code, error)
+        )
+    return r.json()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: wrap this in try-except

File: Integrations/VMRay/integration-VMRay.py [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+SEVERITY_DICT = {
+    "malicious": "Malicious",
+    "suspicious": "Suspicious",
+    "not_suspicious": "Good",
+    "blacklisted": "Blacklisted",
+    "whitelisted": "Whitelisted",
+    "unknown": "Unknown",
+}
+
+DBOTSCORE = {
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: did we consider making `Blacklisted` to get configurable score?

File: Integrations/VMRay/integration-VMRay.py [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+    hashes = ["MD5", "SHA256", "SHA1", "SSDeep"]
+    scores = list()
+    for hash_type in hashes:
+        if hash_type in analysis:
+            scores.append(
+                {
+                    "Indicator": analysis.get(hash_type),
+                    "Type": "hash",
+                    "Vendor": "VMRay",
+                    "Score": DBOTSCORE[analysis.get("Severity")],
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: give default value to `get()`

File: Integrations/VMRay/integration-VMRay.py [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+
+    Returns:
+        dict: response json
+    """
+
+    url = SERVER + url_suffix
+    r = requests.request(
+        method,
+        url,
+        json=body,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: you are not passing `json` from any request function
remove it if not needed

File: Integrations/VMRay/integration-VMRay.py [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+        params["sample_type"] = sample_type
+    if shareable == "true":
+        params["shareable"] = shareable
+    if reanalyze == "true":
+        params["reanalyze"] = reanalyze
+    if max_jobs:
+        if max_jobs.isdigit():
+            params["max_jobs"] = int(max_jobs)
+        else:
+            return_error("max_jobs arguments isn't a number")
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: improve error message

File: Integrations/VMRay/integration-VMRay.py [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+        params["tags"] = tags
+
+    # Request call
+    raw_response = upload_sample(path, params=params)
+    data = raw_response.get("data")
+    jobs_list = list()
+    jobs = data.get("jobs")
+    if jobs:
+        for job in jobs:
+            if isinstance(job, dict):
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: why is this needed? what else can it be? what happens in case it isnt a dict ?

File: Integrations/VMRay/integration-VMRay.py [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+        "Jobs ID": [job.get("JobID") for job in jobs_list],
+        "Samples ID": [sample.get("SampleID") for sample in samples_list],
+        "Submissions ID": [
+            submission.get("SubmissionID") for submission in submissions_list
+        ],
+    }
+    md = tableToMarkdown(
+        "File submitted to VMRay",
+        t=table,
+        headers=["Jobs ID", "Samples ID", "Submissions ID"],
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: add `removeNull=True`

File: Integrations/VMRay/integration-VMRay.py [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+    )
+
+    return_outputs(readable_output=md, outputs=ec, raw_response=raw_response)
+
+
+def build_analysis_data(analyses):
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: `analyses` -> `analysis`

File: Integrations/VMRay/integration-VMRay.py [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+    """
+
+    Args:
+        analyses: (dict) of analysis
+
+    Returns:
+        dict: formatted entry context
+    """
+    ec = dict()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: this is redundant, just initialise the dictionary once with the values

File: Integrations/VMRay/integration-VMRay.py [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+    ec["VMRay.Analysis(val.AnalysisID === obj.AnalysisID)"] = [
+        {
+            "AnalysisID": analysis.get("analysis_id"),
+            "SampleID": analysis.get("analysis_sample_id"),
+            "Severity": SEVERITY_DICT.get(analysis.get("analysis_severity")),
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: add default values here

File: Integrations/VMRay/integration-VMRay.py [hunk 16]
```diff
@@ -0,0 +1,636 @@
...
+    response = http_request("GET", suffix, params=params)
+    return response
+
+
+def get_submission_command():
+    submission_id = demisto.args().get("submission_id")
+    raw_response = get_submission(submission_id)
+    data = raw_response.get("data")
+    # Build entry
+    entry = dict()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: this is redundant, just initialise the dict once with the values

File: Integrations/VMRay/integration-VMRay.py [hunk 17]
```diff
@@ -0,0 +1,636 @@
...
+    entry["IsFinished"] = data.get("submission_finished")
+    entry["HasErrors"] = data.get("submission_has_errors")
+    entry["SubmissionID"] = data.get("submission_id")
+    entry["MD5"] = data.get("submission_sample_md5")
+    entry["SHA1"] = data.get("submission_sample_sha1")
+    entry["SHA256"] = data.get("submission_sample_sha256")
+    entry["SSDeep"] = data.get("submission_sample_ssdeep")
+    entry["Severity"] = SEVERITY_DICT.get(data.get("submission_severity"))
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: add default value to `get`

File: Integrations/VMRay/integration-VMRay.py [hunk 18]
```diff
@@ -0,0 +1,636 @@
...
+    response = http_request("GET", url_suffix=suffix)
+    return response
+
+
+def get_sample_command():
+    sample_id = demisto.args().get("sample_id")
+    raw_response = get_sample(sample_id)
+    data = raw_response.get("data")
+
+    entry = dict()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: this is redundant, just initialise the dict once with the values

File: Integrations/VMRay/integration-VMRay.py [hunk 19]
```diff
@@ -0,0 +1,636 @@
...
+    entry["Classifications"] = data.get("sample_classifications")
+    scores = score_by_hash(entry)
+
+    ec = {
+        "VMRay.Samples(var.SampleID === obj.SampleID)": entry,
+        outputPaths.get("dbotscore"): scores,
+    }
+
+    md = tableToMarkdown(
+        "Results for sample id: {} with severity {}".format(
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: improve title

File: Integrations/VMRay/integration-VMRay.py [hunk 20]
```diff
@@ -0,0 +1,636 @@
...
+    entry["SSDeep"] = data.get("sample_ssdeephash")
+    entry["Severity"] = SEVERITY_DICT.get(data.get("sample_severity"))
+    entry["Type"] = data.get("sample_type")
+    entry["Created"] = data.get("sample_created")
+    entry["Classifications"] = data.get("sample_classifications")
+    scores = score_by_hash(entry)
+
+    ec = {
+        "VMRay.Samples(var.SampleID === obj.SampleID)": entry,
+        outputPaths.get("dbotscore"): scores,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 21]
```diff
@@ -0,0 +1,636 @@
...
+
+    Returns:
+
+    """
+    suffix = "job/sample/{}".format(sample_id)
+    response = http_request("GET", suffix)
+    return response
+
+
+def get_job_sample_command():
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: keep the order of functions consistent, command above request function

File: Integrations/VMRay/integration-VMRay.py [hunk 22]
```diff
@@ -0,0 +1,636 @@
...
+    sample_id = demisto.args().get("sample_id")
+    raw_response = get_job_sample(sample_id)
+    data = raw_response.get("data")
+
+    entry = dict()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: this is redundant, just initialise the dict once with the values

File: Integrations/VMRay/integration-VMRay.py [hunk 23]
```diff
@@ -0,0 +1,636 @@
...
+    entry["SHA1"] = data.get("job_sample_sha1")
+    entry["SHA256"] = data.get("job_sample_sha256")
+    entry["SSDeep"] = data.get("job_sample_ssdeep")
+    entry["JobVMName"] = data.get("job_vm_name")
+    entry["JobVMID"] = data.get("job_vm_id")
+
+    ec = {"VMRay.Jobs(val.JobID === obj.JobID)": entry}
+
+    md = tableToMarkdown(
+        "Results for job sample id: {}".format(sample_id),
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: improve title

File: Integrations/VMRay/integration-VMRay.py [hunk 24]
```diff
@@ -0,0 +1,636 @@
...
+    suffix = "sample/{}/threat_indicators".format(sample_id)
+    response = http_request("GET", suffix).get("data")
+    return response
+
+
+def get_threat_indicators_command():
+    sample_id = demisto.args().get("sample_id")
+    raw_response = get_threat_indicators(sample_id)
+    data = raw_response.get("threat_indicators")
+    # Build Entry Context TODO: EntryContext
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: ??

File: Integrations/VMRay/integration-VMRay.py [hunk 25]
```diff
@@ -0,0 +1,636 @@
...
+            ec_list[0],
+            headers=["AnalysisIDs", "Category", "Classifications", "Operation"],
+        )
+
+        ec = {"VMRay.ThreatIndicators(obj.ID === val.ID)": ec_list}
+        return_outputs(md, ec, raw_response={"threat_indicators": data})
+    return_outputs(
+        "No threat indicators for sample ID: {}".format(sample_id),
+        {},
+        raw_response=raw_response,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 26]
```diff
@@ -0,0 +1,636 @@
...
+            entry = dict()
+            entry["AnalysisIDs"] = indicator.get("analysis_ids")
+            entry["Category"] = indicator.get("category")
+            entry["Classifications"] = indicator.get("classifications")
+            entry["ID"] = indicator.get("id")
+            entry["Operation"] = indicator.get("operation")
+            ec_list.append(entry)
+
+        md = tableToMarkdown(
+            "Threat indicators for sample ID: {}. Showing first indicator:".format(
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: why are we showing only first indicator ?

File: Integrations/VMRay/integration-VMRay.py [hunk 27]
```diff
@@ -0,0 +1,636 @@
...
+
+def post_tags():
+    analysis_id = demisto.args().get("analysis_id")
+    submission_id = demisto.args().get("submission_id")
+    tag = demisto.args().get("tag")
+    if not submission_id and not analysis_id:
+        return_error("No submission ID or analysis ID has been provided")
+    if analysis_id:
+        analysis_status = post_tags_to_analysis(analysis_id, tag)
+        if analysis_status.get("result") == "ok":
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: and if `not` ?

File: Integrations/VMRay/integration-VMRay.py [hunk 28]
```diff
@@ -0,0 +1,636 @@
...
+            return_outputs(
+                "Tags: {} has been added to analysis:".format(tag, analysis_id),
+                {},
+                raw_response=analysis_status,
+            )
+    if submission_id:
+        submission_status = post_tags_to_submission(submission_id, tag)
+        if submission_status.get("result") == "ok":
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: and if `not` ?

File: Integrations/VMRay/integration-VMRay.py [hunk 29]
```diff
@@ -0,0 +1,636 @@
...
+
+def delete_tags():
+    analysis_id = demisto.args().get("analysis_id")
+    submission_id = demisto.args().get("submission_id")
+    tag = demisto.args().get("tags")
+    if not submission_id and not analysis_id:
+        return_error("No submission ID or analysis ID has been provided")
+    if analysis_id:
+        analysis_status = delete_tags_from_analysis(analysis_id, tag)
+        if analysis_status.get("result") == "ok":
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: and if `not` ?

File: Integrations/VMRay/integration-VMRay.py [hunk 30]
```diff
@@ -0,0 +1,636 @@
...
+            return_outputs(
+                "Tags: {} has been added to analysis:".format(tag, analysis_id),
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: deleted?

File: Integrations/VMRay/integration-VMRay.py [hunk 31]
```diff
@@ -0,0 +1,636 @@
...
+                {},
+                raw_response=analysis_status,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 32]
```diff
@@ -0,0 +1,636 @@
...
+            )
+    if submission_id:
+        submission_status = delete_tags_from_submission(submission_id, tag)
+        if submission_status.get("result") == "ok":
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: and if `not` ?

File: Integrations/VMRay/integration-VMRay.py [hunk 33]
```diff
@@ -0,0 +1,636 @@
...
+            return_outputs(
+                "Tags: {} has been added to submission:".format(tag, submission_id),
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: deleted?

File: Integrations/VMRay/integration-VMRay.py [hunk 34]
```diff
@@ -0,0 +1,636 @@
...
+                {},
+                raw_response=submission_status,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: remove trailing comma

File: Integrations/VMRay/integration-VMRay.py [hunk 35]
```diff
@@ -0,0 +1,636 @@
...
+def get_iocs(sample_id):
+    suffix = "sample/{}/iocs".format(sample_id)
+    response = http_request("GET", suffix)
+    return response
+
+
+def get_iocs_command():
+    sample_id = demisto.args().get("sample_id")
+    raw_response = get_iocs(sample_id)
+    data = raw_response.get("data").get("iocs")
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: add default values to `get`

File: Integrations/VMRay/integration-VMRay.py [hunk 36]
```diff
@@ -0,0 +1,636 @@
...
+
+    domains = data.get("domains")
+    domain_list = list()
+    if domains:
+        for domain in domains:
+            entry = dict()
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: this is redundant, just initialise the dictionary once with the values

File: Integrations/VMRay/integration-VMRay.py [hunk 37]
```diff
@@ -0,0 +1,636 @@
...
+    iocs = {
+        "URLs": urls_list,
+        "Mutexes": mutex_list,
+        "Domains": domain_list,
+        "Registry": registry_list,
+        "IPs": ip_list
+    }
+
+    ec = {
+        "VMRay.Samples(val.SampleID == {}).IOCs".format(sample_id): iocs,
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: `===`

File: Integrations/VMRay/integration-VMRay.py [hunk 38]
```diff
@@ -0,0 +1,636 @@
...
+    }
+
+    # Get total size of iocs for HumanReadable
+    iocs_size_table = dict()
+    iocs_size = 0
+    for k, v in iocs.items():
```
Comment (Itay4) [inline Integrations/VMRay/integration-VMRay.py]: you can do that in previous loops

**Comment by jochman** [comment]
> Good job 👍
> In addition to my comments, a few generic ones:
> 
> 1. Make sure @kirbles19 reviews the descriptions, as well as the error messages
> 2. Please post the error messages returned for bad credentials (url, api key..)
> 3. Context outputs should be improved: use singular (for example `Jobs`->`Job`) and use generic context standards (`File`, `Domain`, etc...), add `Malicious` context when needed
> 4. Use `'` instead of `"` in all code.
> 5. Don't spare in variable names (`md` -> `human_readable`, `ec` -> `entry_context`
> 6. Do we return DBotScore 0 when a file is not found ?
> 7. Deprecate old commands
> 8. Why is `Integrations/__init__.py` in changed files?
> 9. Please post playbooks screenshots

about 4:
since we're using flake8/pep8, we should use `"` instead of `'`.

6. we do return dbot score of 0.
9. next comment

**Comment by jochman** [comment]
@kirbles19 please review yml https://github.com/demisto/content/blob/3a24c1f3f606df332e0b85263607265c12e8e346/Integrations/VMRay/VMRay.yml

**Comment by jochman** [comment]
image

**Comment by jochman** [comment]
https://github.com/demisto/etc/issues/16816 other connected issue
